*** dbus-1.3.0/bus/system.conf.in	2010-05-10 11:53:56.792067068 -0700
--- dbus-1.3.0.new/bus/system.conf.in	2010-05-10 11:44:09.184062000 -0700
***************
*** 47,54 ****
  
      <!-- Holes must be punched in service configuration files for
           name ownership and sending method calls -->
!     <deny own="*"/>
!     <deny send_type="method_call"/>
  
      <!-- Signals and reply messages (method returns, errors) are allowed
           by default -->
--- 47,54 ----
  
      <!-- Holes must be punched in service configuration files for
           name ownership and sending method calls -->
!     <allow own="*"/>
!     <allow send_type="method_call"/>
  
      <!-- Signals and reply messages (method returns, errors) are allowed
           by default -->
