Index: b/usbmount.conf
===================================================================
--- a/usbmount.conf
+++ b/usbmount.conf
@@ -14,7 +14,7 @@ MOUNTPOINTS="/media/usb0 /media/usb1 /me
 
 # Filesystem types: removable storage devices are only mounted if they
 # contain a filesystem type which is in this list.
-FILESYSTEMS="vfat ext2 ext3 ext4 hfsplus"
+FILESYSTEMS="vfat ext2 ext3 ext4 xfs exfat ntfs hfsplus"
 
 #############################################################################
 # WARNING!                                                                  #
