From 00bc8e3031dd5c4d9f33c2a92b8eac8b235940d7 Mon Sep 17 00:00:00 2001
From: Trent Nelson <trent.nelson@pivosgroup.com>
Date: Fri, 18 Oct 2013 23:43:54 -0400
Subject: [PATCH 1/4] if devinfo is empty, ensure variable is empty

---
 usbmount | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/usbmount b/usbmount
index 75f249a..f887f68 100755
--- a/usbmount
+++ b/usbmount
@@ -88,9 +88,9 @@ if [ "$1" = add ]; then
     #   FIXME: improvement: implement mounting by label (notice that labels
     #   can contain spaces, which makes things a little bit less comfortable).
     DEVINFO=$(/sbin/blkid -p $DEVNAME)
-    FSTYPE=$(echo "$DEVINFO" | sed 's/.*[[:blank:]]TYPE="\([^"]*\)".*/\1/g; s/[[:blank:]]*//g;')
-    UUID=$(echo "$DEVINFO"   | sed 's/.*[[:blank:]]UUID="\([^"]*\)".*/\1/g; s/[[:blank:]]*//g;')
-    USAGE=$(echo "$DEVINFO"  | sed 's/.*[[:blank:]]USAGE="\([^"]*\)".*/\1/g; s/[[:blank:]]*//g;')
+    FSTYPE=$(echo "$DEVINFO" | sed -ne 's/.*[[:blank:]]TYPE="\([^"]*\)".*/\1/pg; s/[[:blank:]]*//g;')
+    UUID=$(echo "$DEVINFO"   | sed -ne 's/.*[[:blank:]]UUID="\([^"]*\)".*/\1/pg; s/[[:blank:]]*//g;')
+    USAGE=$(echo "$DEVINFO"  | sed -ne 's/.*[[:blank:]]USAGE="\([^"]*\)".*/\1/pg; s/[[:blank:]]*//g;')
 
     if ! echo $USAGE | egrep -q "(filesystem|disklabel)"; then
 	log info "$DEVNAME does not contain a filesystem or disklabel"
-- 
1.8.1.2

